### ISSUE TITLE
実装: 競りMonte Carlo Q Network trainerとQ値検証を追加する

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#370]
Related: #202, #367, #368
Base branch: `feature/issue-202-non-playing-ai`

## 背景

Issue #367 で、競り学習について現行Actor-Critic PPOよりも、各legal bidding actionの期待終局価値 `Q(s,a)` を直接学習する **Monte Carlo fitted Q regression with counterfactual action coverage** を有力候補とした。

Issue #368 では、その教師データ生成基盤を実装した。

現在、以下の形式のraw Q教師datasetを生成できる。

- compact278 model input
- 29-entry legal bid mask
- forced action index / semantic action
- strongest suit / score
- repeat index / rollout seed
- bidding-specific terminal reward
- terminal role/result / contract success
- state / artifact / opponent provenance

smokeでは100 source statesから661 forced state-action samplesを生成し、PASS / S / H / D / Cおよびaction index 0..28を全てcoverageできた。

次は、このdatasetから29-output Q Networkを学習し、counterfactual teacher valueをactionごとに再現・順位付けできるかを検証する。

## 目的

競り専用のMonte Carlo Q Network trainerを追加する。

初版network:

```text
compact278 -> MLP -> 29 Q values
```

各sample `(s,a,target)` について、選択されたaction `a` の出力 `Q(s,a)` だけをMonte Carlo terminal targetへ回帰する。

このIssueではruntime Q bidding agentを本番対戦へ接続しない。
まずoffline / held-out validationでQ値学習そのものが成立することを確認する。

## 基本アルゴリズム

名称:

`Monte Carlo fitted Q regression`

target:

```text
Q_target(s,a) = mean terminal bidding reward for repeated counterfactual rollouts of (s,a)
```

Bellman bootstrap / target network / DQNは初版では使用しない。

## Dataset input

Issue #368 の dataset schema v1 を正式入力とする。

trainerはraw JSONL samplesを読み込み、少なくとも以下でgroupできること。

```text
(stateKey, forcedActionIndex)
```

同一pairに複数repeatがある場合:

- target mean
- target variance
- repeat count

を計算する。

raw sampleを破壊せず、trainer内部またはtraining reportで集約する。

## Train / validation split

state leakageを防ぐため、sample単位random splitではなく **stateKey単位** でtrain/validationを分ける。

同一stateの異なるcounterfactual actionsがtrainとvalidationへ跨がらないこと。

seed固定で再現可能にする。

推奨初期split:

- train 80%
- validation 20%

CLIで変更可能でもよい。

## Q Network

初版は29-output MLP。

### input

- compact278

### output

- 29 raw Q values

action mappingは既存と完全一致:

- 0 = PASS
- 1..28 = target 13..19 x S/H/D/C

### architecture

まず既存bidding MLPと比較しやすいhidden dimsをdefault候補とする。

例:

```text
[512, 512, 256, 256]
```

ただしQ modelはPPO actor checkpointから初期化せずfresh initializationとする。

model / checkpoint metadataに明確なarchitecture idを持たせる。

例:

`bidding-q-mlp-v1`

## Loss

初版はHuber lossを推奨する。

各aggregated `(state, action)` sampleについて:

```text
prediction = Q(state)[actionIndex]
loss = huber(prediction, meanTerminalReward)
```

教師が存在しない他28 outputsへlossを掛けない。

比較用にMSEをoptionとして持たせてもよいがdefaultはHuberとする。

## Sample weighting / balancing

PPOでaction disappearance / suit collapseが起きたため、Q trainerでもaction coverageを明示的に扱う。

最低限、training reportで以下を出す。

- PASS sample count
- bid sample count
- S/H/D/C sample count
- target 13..19 sample count
- action index 0..28 count
- strongest suit count

trainer minibatchは、単純randomをdefaultにしてよいが、少数actionが実質無視されないことを確認する。

可能なら初版で以下の軽量balanced samplerを追加する。

- PASS / bid family
- S/H/D/C family

ただし人工的oversamplingがvalidation metricを歪めないよう、validationはraw distributionで行う。

trainerが複雑化する場合は、このIssueではまずrandom + per-action diagnosticsまででもよい。その場合、coverageが十分なdatasetを使って検証する。

## Target normalization

raw reward scaleは、winで `2*d` / `3*d`、loss `-1`、non-contract `0` と非対称。

以下を比較し、実装として安全な方式を選ぶ。

- raw targetのまま
- train splitのmean/stdでstandardize

初版はraw targetでもよいが、学習が不安定ならnormalizationをoption化する。

normalizationする場合:

- train statisticsのみで算出
- metadataへmean/std保存
- validation/reportは必ずraw Q scaleへ戻して表示

## CLI

新しいtrainer CLIを追加する。

最低限指定可能:

- dataset dir
- output dir
- epochs
- batch size
- learning rate
- hidden dims
- loss type
- train/validation split
- seed
- training device

命名例:

```text
napoleon-train-bidding-q
```

既存Python CLI規約に合わせて調整してよい。

## Checkpoint / artifact

最低限保存:

- checkpoint.pt
- model metadata JSON
- training config
- dataset manifest SHA / source dataset identity
- architecture id
- compact278 schema/version
- 29-action mapping id
- reward id/version
- loss / optimizer config
- split seed
- train/validation state counts
- training metrics

旧PPO artifactと誤認しないpolicy/model typeにする。

## ONNX export

29-output Q NetworkをONNX exportできるようにする。

shape:

```text
(batch, 278) -> (batch, 29)
```

metadataでpolicy logitsではなくQ valuesであることを識別できるようにする。

例:

- modelType / policyType: bidding-q
- selectionSemantics: legal-q-argmax

既存runtime agentへ接続は今回非目標だが、後続Issueでload可能なartifact形式にする。

## Validation metrics

このIssueの中心。

単純lossだけでなく、Qがaction価値を学べているかを測る。

### Regression

validation setで:

- Huber / MSE
- MAE
- RMSE
- predicted Q mean/std
- target mean/std
- Pearson correlation（可能なら）

### Calibration

- predicted Q decileごとのactual mean target
- predicted vs actual scatter用データ
- action family別 MAE
  - PASS
  - S/H/D/C
  - target 13..19

### Ranking

同一validation stateに複数counterfactual action teacherがある場合:

- empirical best action = highest teacher mean
- predicted best action = highest predicted Q among teacher-covered legal actions

最低限:

- best-action hit rate
- top-3 hit rate
- pairwise ranking accuracy
- predicted top1 - top2 Q gap

さらに:

- PASS vs best-bid ranking accuracy
- strongest suit bid vs other suit ranking
- suit-level best hit rate
- target-level best hit rate

を可能な範囲で出す。

## Important: teacher-covered action mask

validation rankingでは、teacherが無いactionまで含めて「正解best action」と比較しない。

各stateでcounterfactual teacherが存在するaction subset内でranking metricを計算する。

full 29-action rankingは、後続のheld-out exhaustive counterfactual evaluationで行う。

## Q-value diagnostics

validation stateごとに可能なら以下を出す。

- PASS Q
- best teacher-covered bid Q
- PASS - bestBid Q gap
- predicted selected action
- strongest suit
- selected suit
- strongest suit match

aggregate:

- argmax PASS rate on teacher-covered set
- S/H/D/C predicted best distribution
- strongest suit match rate
- target predicted distribution

これによりQ networkが再びPASSまたは13-Cへcollapseしていないか確認する。

## Small training experiment

実装後、Issue #368 generatorで、smokeより少し大きいdatasetを生成してQ trainerを実際に学習させる。

推奨:

- 500〜1000 source states程度
- candidate rotation
- 5〜8 forced actions/state
- repeats=2程度（runtimeが許せば）
- reward: `bidding-q-contract-result-loss-minus-one-v1`

既存smoke dataset 661 samplesだけでもtrainer smokeには使えるが、性能判断はより大きいdatasetを優先する。

学習後、train/validation metricsとranking diagnosticsをIssueへ記録する。

## Success criteria for first Q trainer

完璧なゲーム性能はまだ要求しない。

最低限、以下を満たせるかを見る。

- train lossが低下する
- validation MAEが単純baselineを上回る
- predicted Qとteacher targetに正の相関がある
- best-action hit rateがrandom baselineを明確に上回る
- PASS/BIDだけでなくsuit rankingにもsignalがある
- S/H/D/Cのpredicted best actionが全て少なくとも一定数存在する、または欠落理由をteacher data側から説明できる
- fixed strongest-suit subsetでQ rankingが完全ランダムではない

## Baselines

最低限、単純baselineと比較する。

候補:

1. global mean reward
2. action-index mean reward
3. suit/target bucket mean reward

Q Networkがこれらよりvalidation MAE / rankingで改善することを確認する。

特にaction-index mean baselineは重要。NNがstate情報を使っているかを見る。

## Tests

最低限:

1. Issue #368 dataset schema loader
2. stateKey単位splitでleakしない
3. repeat aggregation mean/variance/count
4. 278 -> 29 forward shape
5. selected action only loss
6. Huber loss fixture
7. illegal / teacher-missing actionをloss対象にしない
8. deterministic seed training smoke
9. checkpoint save/load
10. metadata / dataset provenance validation
11. ONNX export shape parity
12. PyTorch vs ONNX Q output parity
13. validation ranking metric fixture
14. action/suit/target diagnostics correctness
15. CPU training smoke
16. CUDA利用可能環境ではCUDA smoke（可能なら）

## Non-goals

- runtime Q bidding agent接続
- Q policyで実ゲームevaluation
- iterative replay buffer loop
- epsilon-greedy online collection
- Bellman/DQN target
- target network
- PPO削除
- existing PPO artifact変更
- adjutant/exchange/playing再学習
- mainへのMerge

## Git / PR運用

最新 `feature/issue-202-non-playing-ai` からIssue専用ブランチを作成する。

PR baseは必ず `feature/issue-202-non-playing-ai`。
main向けPRは作らない。

実装・unit/integration tests・small Q training experiment完了後、Codex review最大4回（review後修正最大3回）を実施する。

blocking / actionable指摘がなく、必要test・experimentが成功したらReady for reviewにし、親ブランチへmergeする。

mainにはmergeしない。

## Acceptance Criteria

- Issue #368 datasetを読み込める
- `(stateKey, action)` repeatをMonte Carlo mean targetへ集約できる
- stateKey単位train/validation splitがある
- compact278 -> 29 Q values modelがある
- selected/forced actionだけHuber regressionできる
- fresh Q checkpointを学習できる
- Q artifact/provenanceがPPOと明確に区別される
- ONNX exportできる
- PyTorch / ONNX output parity確認
- validation MAE/RMSE/correlationを出せる
- best-action hit / top3 / pairwise ranking metricsを出せる
- PASS/S/H/D/C/target別diagnosticsを出せる
- global/action-mean等baselineと比較できる
- small training experimentでend-to-end完走する
- Qがstate-conditioned action valueを学べているか具体的結論をIssueへ残す
